File: examples/python/part-6/18-inter-agent-security.py
# Illustrative examples for notes/part-6-multi-agent-security/18-inter-agent-security.md
# Not for production use. Licensed under MIT (see LICENSE-CODE).

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Dict, List, Protocol


AgentID = str


class Scope(str, Enum):
    READ_DOCS = "read:docs"
    WEB_SEARCH = "web:search"
    SEND_EMAIL = "send:email"
    DB_WRITE = "db:write"
    SHELL_RUN = "shell:run"


@dataclass
class AgentIdentity:
    id: AgentID
    role: str
    allowed_tools: List[str] = field(default_factory=list)
    allowed_scopes: List[Scope] = field(default_factory=list)
    handoff_targets: List[AgentID] = field(default_factory=list)
    prompt_version: str = ""
    policy_version: str = ""


def has_scope(scopes: List[Scope], want: Scope) -> bool:
    return want in scopes


class TrustLevel(str, Enum):
    TRUSTED_RUNTIME = "trusted_runtime"
    UNTRUSTED_INPUT = "untrusted_input"
    AGENT_OUTPUT = "agent_output"


@dataclass
class AgentMessage:
    id: str
    run_id: str
    parent_action_id: str
    from_agent: AgentID
    to_agent: AgentID
    task: str
    data_refs: List[str] = field(default_factory=list)
    delegated_scopes: List[Scope] = field(default_factory=list)
    trust: TrustLevel = TrustLevel.TRUSTED_RUNTIME
    created_at: datetime = field(default_factory=datetime.utcnow)


@dataclass
class HandoffPolicy:
    agents: Dict[AgentID, AgentIdentity]

    def allow_handoff(self, msg: AgentMessage) -> None:
        from_agent = self.agents.get(msg.from_agent)
        if from_agent is None:
            raise ValueError(f"unknown source agent: {msg.from_agent}")

        to_agent = self.agents.get(msg.to_agent)
        if to_agent is None:
            raise ValueError(f"unknown target agent: {msg.to_agent}")

        if not _contains_agent(from_agent.handoff_targets, msg.to_agent):
            raise ValueError(f"handoff from {msg.from_agent} to {msg.to_agent} is not allowed")

        for scope in msg.delegated_scopes:
            if not has_scope(from_agent.allowed_scopes, scope):
                raise ValueError(f"source agent does not own scope: {scope.value}")
            if not has_scope(to_agent.allowed_scopes, scope):
                raise ValueError(f"target agent cannot receive scope: {scope.value}")

        if msg.trust == TrustLevel.UNTRUSTED_INPUT:
            raise ValueError("untrusted input cannot be delegated without sanitization")


def _contains_agent(items: List[AgentID], want: AgentID) -> bool:
    return want in items


@dataclass
class HandoffBudget:
    max_depth: int
    max_handoffs: int
    max_agents: int
    depth: int = 0
    handoffs: int = 0
    agents_involved: Dict[AgentID, bool] = field(default_factory=dict)

    def check(self, next_agent: AgentID) -> None:
        self.handoffs += 1
        self.agents_involved[next_agent] = True

        if self.depth > self.max_depth:
            raise RuntimeError("max handoff depth exceeded")
        if self.handoffs > self.max_handoffs:
            raise RuntimeError("max handoffs exceeded")
        if len(self.agents_involved) > self.max_agents:
            raise RuntimeError("max agents involved exceeded")


class Agent(Protocol):
    def run(self, msg: AgentMessage) -> AgentMessage: ...


class AuditLogger(Protocol):
    def log_handoff(self, msg: AgentMessage, decision: str, reason: str) -> None: ...


@dataclass
class HandoffExecutor:
    policy: HandoffPolicy
    agents: Dict[AgentID, Agent]
    audit: AuditLogger
    budget: HandoffBudget

    def execute(self, msg: AgentMessage) -> AgentMessage:
        try:
            self.budget.check(msg.to_agent)
        except Exception as exc:
            self.audit.log_handoff(msg, "denied", str(exc))
            raise

        try:
            self.policy.allow_handoff(msg)
        except Exception as exc:
            self.audit.log_handoff(msg, "denied", str(exc))
            raise

        agent = self.agents.get(msg.to_agent)
        if agent is None:
            reason = f"agent not registered: {msg.to_agent}"
            self.audit.log_handoff(msg, "denied", reason)
            raise ValueError(reason)

        self.audit.log_handoff(msg, "allowed", "handoff policy passed")
        return agent.run(msg)

File: examples/typescript/part-6/18-inter-agent-security.ts
// Illustrative examples for notes/part-6-multi-agent-security/18-inter-agent-security.md
// Not for production use. Licensed under MIT (see LICENSE-CODE).

type AgentID = string;

enum Scope {
  ReadDocs = "read:docs",
  WebSearch = "web:search",
  SendEmail = "send:email",
  DBWrite = "db:write",
  ShellRun = "shell:run",
}

interface AgentIdentity {
  id: AgentID;
  role: string;
  allowedTools: string[];
  allowedScopes: Scope[];
  handoffTargets: AgentID[];
  promptVersion: string;
  policyVersion: string;
}

function hasScope(scopes: Scope[], want: Scope): boolean {
  return scopes.includes(want);
}

enum TrustLevel {
  TrustedRuntime = "trusted_runtime",
  UntrustedInput = "untrusted_input",
  AgentOutput = "agent_output",
}

interface AgentMessage {
  id: string;
  runId: string;
  parentActionId: string;
  from: AgentID;
  to: AgentID;
  task: string;
  dataRefs: string[];
  delegatedScopes: Scope[];
  trust: TrustLevel;
  createdAt: Date;
}

class HandoffPolicy {
  agents: Record<AgentID, AgentIdentity>;

  constructor(agents: Record<AgentID, AgentIdentity>) {
    this.agents = agents;
  }

  allowHandoff(msg: AgentMessage): void {
    const from = this.agents[msg.from];
    if (!from) {
      throw new Error(`unknown source agent: ${msg.from}`);
    }

    const to = this.agents[msg.to];
    if (!to) {
      throw new Error(`unknown target agent: ${msg.to}`);
    }

    if (!containsAgent(from.handoffTargets, msg.to)) {
      throw new Error(`handoff from ${msg.from} to ${msg.to} is not allowed`);
    }

    for (const scope of msg.delegatedScopes) {
      if (!hasScope(from.allowedScopes, scope)) {
        throw new Error(`source agent does not own scope: ${scope}`);
      }
      if (!hasScope(to.allowedScopes, scope)) {
        throw new Error(`target agent cannot receive scope: ${scope}`);
      }
    }

    if (msg.trust === TrustLevel.UntrustedInput) {
      throw new Error("untrusted input cannot be delegated without sanitization");
    }
  }
}

function containsAgent(items: AgentID[], want: AgentID): boolean {
  return items.includes(want);
}

class HandoffBudget {
  maxDepth: number;
  maxHandoffs: number;
  maxAgents: number;
  depth = 0;
  handoffs = 0;
  agentsInvolved: Record<AgentID, boolean> = {};

  constructor(maxDepth: number, maxHandoffs: number, maxAgents: number) {
    this.maxDepth = maxDepth;
    this.maxHandoffs = maxHandoffs;
    this.maxAgents = maxAgents;
  }

  check(next: AgentID): void {
    this.handoffs += 1;
    this.agentsInvolved[next] = true;

    if (this.depth > this.maxDepth) {
      throw new Error("max handoff depth exceeded");
    }
    if (this.handoffs > this.maxHandoffs) {
      throw new Error("max handoffs exceeded");
    }
    if (Object.keys(this.agentsInvolved).length > this.maxAgents) {
      throw new Error("max agents involved exceeded");
    }
  }
}

interface Agent {
  run(msg: AgentMessage): AgentMessage;
}

interface AuditLogger {
  logHandoff(msg: AgentMessage, decision: string, reason: string): void;
}

class HandoffExecutor {
  policy: HandoffPolicy;
  agents: Record<AgentID, Agent>;
  audit: AuditLogger;
  budget: HandoffBudget;

  constructor(
    policy: HandoffPolicy,
    agents: Record<AgentID, Agent>,
    audit: AuditLogger,
    budget: HandoffBudget,
  ) {
    this.policy = policy;
    this.agents = agents;
    this.audit = audit;
    this.budget = budget;
  }

  execute(msg: AgentMessage): AgentMessage {
    try {
      this.budget.check(msg.to);
    } catch (err) {
      this.audit.logHandoff(msg, "denied", String(err));
      throw err;
    }

    try {
      this.policy.allowHandoff(msg);
    } catch (err) {
      this.audit.logHandoff(msg, "denied", String(err));
      throw err;
    }

    const agent = this.agents[msg.to];
    if (!agent) {
      const reason = `agent not registered: ${msg.to}`;
      this.audit.logHandoff(msg, "denied", reason);
      throw new Error(reason);
    }

    this.audit.logHandoff(msg, "allowed", "handoff policy passed");
    return agent.run(msg);
  }
}
